account-integrations/safe/test/hardhat/utils/createUserOp.ts

```diff
@@ -0,0 +1,169 @@
+import { ethers, getBytes, HDNodeWallet, NonceManager } from "ethers";
+import { AddressZero } from "@ethersproject/constants";
+import { UserOperationStruct } from "@account-abstraction/contracts";
+import { getUserOpHash } from "@account-abstraction/utils";
+
+import { SafeProxyFactory } from "../../../typechain-types/lib/safe-contracts/contracts/proxies/SafeProxyFactory";
+import { Safe } from "../../../typechain-types/lib/safe-contracts/contracts/Safe";
+import {
+  EntryPoint__factory,
+  SafeBlsPlugin,
+  SafeWebAuthnPlugin,
+} from "../../../typechain-types";
+import receiptOf from "./receiptOf";
+import { calculateProxyAddress } from "./calculateProxyAddress";
+import { getGasEstimates } from "./getGasEstimates";
+import sendUserOpAndWait from "./sendUserOpAndWait";
+
+type Plugin = SafeBlsPlugin | SafeWebAuthnPlugin;
+
+export const generateInitCodeAndAddress = async (
+  admin: NonceManager,
+  owner: ethers.HDNodeWallet,
+  plugin: Plugin,
+  safeSingleton: Safe,
+  safeProxyFactory: SafeProxyFactory,
+) => {
+  const pluginAddress = await plugin.getAddress();
+  const singletonAddress = await safeSingleton.getAddress();
+  const factoryAddress = await safeProxyFactory.getAddress();
+
+  const moduleInitializer = plugin.interface.encodeFunctionData("enableMyself");
+  const encodedInitializer = safeSingleton.interface.encodeFunctionData(
+    "setup",
+    [
+      [owner.address],
+      1,
+      pluginAddress,
+      moduleInitializer,
+      pluginAddress,
+      AddressZero,
+      0,
+      AddressZero,
+    ],
+  );
+
+  const deployedAddress = await calculateProxyAddress(
+    safeProxyFactory,
+    await safeSingleton.getAddress(),
+    encodedInitializer,
+    73,
+  );
+
+  // Native tokens for the pre-fund
+  await receiptOf(
+    admin.sendTransaction({
+      to: deployedAddress,
+      value: ethers.parseEther("10"),
+    }),
+  );
+
+  // The initCode contains 20 bytes of the factory address and the rest is the
+  // calldata to be forwarded
+  const initCode = ethers.concat([
+    factoryAddress,
+    safeProxyFactory.interface.encodeFunctionData("createProxyWithNonce", [
+      singletonAddress,
+      encodedInitializer,
+      73,
+    ]),
+  ]);
+
+  return { initCode, deployedAddress };
+};
```

Comment:
Is there a reason you went this route vs. using the contract factory pattern that is used for non-recovery plugins?

Reply:
The reason this route is done is because the original safe 4337 [example test ](https://github.com/safe-global/safe-contracts/blob/main/test/integration/Safe.ERC4337.spec.ts) does something like this. So these plugins haven't been updated to use a factory

At the moment there is no shared way that we're creating these plugins, some use a dedicated factory and others use this method.

We probably need to decide how we want to deploy these plugins. We could treat them as 4337 accounts - there is a pattern of using a contract factory for deploying 4337 accounts (e.g SimpleAccount). Although if the plugins are singletons (which some of them are), then deploying them from a factory might not be appropriate